src/Framework/Framework/Compilation/ControlTree/UnsupportedCallSiteCheckingVisitor.cs

```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using DotVVM.Framework.CodeAnalysis;
+using DotVVM.Framework.Binding.Expressions;
+using DotVVM.Framework.Compilation.ControlTree.Resolved;
+
+namespace DotVVM.Framework.Compilation.ControlTree
+{
+    public class UnsupportedCallSiteCheckingVisitor : ResolvedControlTreeVisitor
+    {
+        class ExpressionInspectingVisitor : ExpressionVisitor
+        {
+            public event Action<MethodInfo>? InvalidCallSiteDetected;
+
+            protected override Expression VisitMethodCall(MethodCallExpression node)
+            {
+                var callSiteAttr = node.Method.CustomAttributes.FirstOrDefault(a => a.AttributeType == typeof(UnsupportedCallSiteAttribute));
+                if (callSiteAttr is not null && callSiteAttr.ConstructorArguments.Any())
```

Comment:
I'd maybe check it using `node.Method.IsDefined(typeof(UnsupportedCallSiteAttribute))`, it's significantly faster and this will be called for all methods in all bindings. Easier to "optimize it" than measure the impact on startup time...